plugins/thunderbolt/fu-thunderbolt-image-validation.c

```diff
@@ -0,0 +1,482 @@
+		{ .offset = 0X11CB, .len = 1, .description = "Snk1_2(DP-in)" },
+		{ .offset = 0X11CF, .len = 1, .description = "Snk1_3(DP-in)" },
+		{ 0 }
+	};
+
+	static const FwLocation WR[] = {
+		{ .offset = 0x10,   .len = 4, .description = "PCIe Settings" },
+		{ .offset = 0x14F,  .len = 1, .description = "CIO-Port0_TX"  },
+		{ .offset = 0x157,  .len = 1, .description = "CIO-Port0_RX"  },
+		{ .offset = 0x153,  .len = 1, .description = "CIO-Port1_TX"  },
+		{ .offset = 0x15B,  .len = 1, .description = "CIO-Port1_RX"  },
+		{ .offset = 0x1F1,  .len = 1, .description = "Snk0_0(DP-in)" },
+		{ .offset = 0x1F5,  .len = 1, .description = "Snk0_1(DP-in)" },
+		{ .offset = 0x1F9,  .len = 1, .description = "Snk0_2(DP-in)" },
+		{ .offset = 0x1FD,  .len = 1, .description = "Snk0_3(DP-in)" },
+		{ .offset = 0X11A5, .len = 1, .description = "PA(DP-out)"    },
+		{ 0 }
+	};
+
+	static const FwLocation AR[] = {
+		{ .offset = 0x10,  .len = 4, .description = "PCIe Settings" },
+		{ .offset = 0x12,  .len = 1, .description = "PA", .mask = 0xCC, .section = DRAM_UCODE_SECTION },
+		{ .offset = 0x121, .len = 1, .description = "Snk0" },
+		{ .offset = 0x129, .len = 1, .description = "Snk1" },
+		{ .offset = 0x136, .len = 1, .description = "Src0", .mask = 0xF0 },
+		{ .offset = 0xB6,  .len = 1, .description = "PA/PB (USB2)", .mask = 0xC0 },
+		{ 0 },
+
+		{ .offset = 0x13, .len = 1, .description = "PB", .mask = 0xCC, .section = DRAM_UCODE_SECTION },
+		{ 0 }
+	};
+
+	static const FwLocation AR_LP[] = {
+		{ .offset = 0x10,  .len = 4, .description = "PCIe Settings" },
+		{ .offset = 0x12,  .len = 1, .description = "PA", .mask = 0xCC, .section = DRAM_UCODE_SECTION },
+		{ .offset = 0x13,  .len = 1, .description = "PB", .mask = 0x44, .section = DRAM_UCODE_SECTION },
+		{ .offset = 0x121, .len = 1, .description = "Snk0" },
+		{ .offset = 0xB6,  .len = 1, .description = "PA/PB (USB2)", .mask = 0xC0 },
+		{ 0 }
+	};
+
+	switch (id)
+	{
+	case 0x156D:
+	case 0x156B:
+		return FR;
+	case 0x157E:
+		return WR;
+	case 0x1578:
+	case 0x1576:
+	case 0x15D3:
+	case 0x15DA:
+		return AR;
+	case 0x15C0:
+		return AR_LP;
+	default:
+		return NULL;
+	}
+}
+
+static FwLocation *
+get_device_locations (guint16 id) {
+	static const FwLocation locations[] = {
+		{ .offset = 0x124, .len = 1, .section = ARC_PARAMS_SECTION, .description = "X of N" },
+		{ 0 }
+	};
+
+	switch (id)
+	{
+	case 0x1578:
+	case 0x1576:
+	case 0x15D3:
+	case 0x15DA:
+	case 0x15C0:
+		return locations;
+	default:
+		return NULL;
+	}
+}
+
+/*
+ * Compares the given locations, assuming locations is an array.
+ * Returns FALSE and sets error upon failure.
+ * locations points to the end of the array (the empty entry) upon
+ * successful return.
+ */
+static gboolean
+compare_locations (const FwLocation **locations,
+		   const FwObject    *controller,
+		   const FwObject    *image,
+		   GError           **error)
+{
+	for ( ; (*locations)->offset != 0; (*locations)++)
```

Comment:
Would a `do {} while () be better here?`

Reply:
You mean something like `while (++(*locations))->offset != 0);`?